src/tiled/changeimagelayerproperties.cpp

```diff
@@ -32,41 +32,87 @@ using namespace Tiled;
 ChangeImageLayerProperties::ChangeImageLayerProperties(
         MapDocument *mapDocument,
         ImageLayer *imageLayer,
-        const QColor &color,
-        const QUrl &newSource)
-    : QUndoCommand(
-          QCoreApplication::translate(
-              "Undo Commands", "Change Image Layer Properties"))
+        const QColor transparentColor)
+    : QUndoCommand(QCoreApplication::translate("Undo Commands",
+                                               "Change Image Layer Transparent Color"))
     , mMapDocument(mapDocument)
     , mImageLayer(imageLayer)
-    , mUndoColor(imageLayer->transparentColor())
-    , mRedoColor(color)
-    , mUndoSource(imageLayer->imageSource())
-    , mRedoSource(newSource)
+    , mProperty(TransparentColorProperty)
+    , mTransparentColor(transparentColor)
 {
 }
 
-void ChangeImageLayerProperties::redo()
+ChangeImageLayerProperties::ChangeImageLayerProperties(
+        MapDocument *mapDocument,
+        ImageLayer *imageLayer,
+        const QUrl imageSource)
+    : QUndoCommand(QCoreApplication::translate("Undo Commands",
+                                               "Change Image Layer Image Source"))
+    , mMapDocument(mapDocument)
+    , mImageLayer(imageLayer)
+    , mProperty(ImageSourceProperty)
+    , mImageSource(imageSource)
 {
-    mImageLayer->setTransparentColor(mRedoColor);
+}
 
-    if (mRedoSource.isEmpty())
-        mImageLayer->resetImage();
-    else
-        mImageLayer->loadFromImage(mRedoSource);
+ChangeImageLayerProperties::ChangeImageLayerProperties(
+        MapDocument *mapDocument,
+        ImageLayer *imageLayer,
+        ChangeImageLayerProperties::Property property,
+        bool repeat)
+    : QUndoCommand(QCoreApplication::translate("Undo Commands",
+                                               "Change Image Layer Repeat Property"))
+    , mMapDocument(mapDocument)
+    , mImageLayer(imageLayer)
+    , mProperty(property)
+    , mRepeat(repeat)
+{
+}
 
-    emit mMapDocument->imageLayerChanged(mImageLayer);
+void ChangeImageLayerProperties::redo()
+{
+    swap();
 }
 
 void ChangeImageLayerProperties::undo()
 {
-    mImageLayer->setTransparentColor(mUndoColor);
+    swap();
+}
+
+void ChangeImageLayerProperties::swap()
+{
+    switch (mProperty) {
+    case TransparentColorProperty: {
+        const QColor color = mImageLayer->transparentColor();
+        mImageLayer->setTransparentColor(mTransparentColor);
+        mTransparentColor = color;
+        break;
+    }
+    case ImageSourceProperty: {
+        const QUrl source = mImageLayer->imageSource();
+        mImageLayer->setSource(mImageSource);
+        mImageSource = source;
 
-    if (mUndoSource.isEmpty())
-        mImageLayer->resetImage();
-    else
-        mImageLayer->loadFromImage(mUndoSource);
+        if (mImageSource.isEmpty())
```

Comment:
Hmm, there's a bug here in that this checks the old `mImageSource`. I was testing the change and when I tried to set the image on the layer it would reset the image instead of load the one I just set. This block should happen before assigning `source` to `mImageSource` (and since `loadFromImage` sets the image URL as well, we can actually just remove the call to `setSource`).